search_agg_date_histogram.go

```diff
@@ -0,0 +1,236 @@
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// DateHistogramAggregation buckets documents based on a date interval.
+// An empty DateHistogramAggregation will have several issues with execution:
+//   - the target Field must be non-null and non-empty
+//   - the Interval must be non-null and non-empty
+//
+// For more details see https://opensearch.org/docs/latest/opensearch/bucket-agg/#histogram-date_histogram
+type DateHistogramAggregation struct {
+	// Field to be bucketed
+	Field string
+
+	// MinDocCount is the lower count threshold for a bucket to be included in the results.
+	// Negative counts will be omitted
+	MinDocCount int64
+
+	// Interval string using OpenSearch [date math].
+	// [date math]: https://opensearch.org/docs/latest/opensearch/supported-field-types/date/#date-math
+	Interval string
+
+	// TimeZone, times are stored internally in UTC and by default date histograms are bucketed in UTC.
+	// Set the TimeZone to overwrite this default
+	TimeZone string
+
+	// Order list of [Order]s to sort the aggregation buckets. Default order is _count: desc
+	Order []Order
+
+	// Aggregations sub aggregations for each bucket. Mapped by string label to sub aggregation
+	Aggregations map[string]Aggregation
+}
+
+// NewDateHistogramAggregation instantiates a DateHistogramAggregation targeting
+// the provided field with the provided interval. Sets the MinDocCount to -1 to be
+// omitted in favor of the OpenSearch default.
+func NewDateHistogramAggregation(field, interval string) *DateHistogramAggregation {
+	return &DateHistogramAggregation{
+		Field:        field,
+		MinDocCount:  -1,
+		Interval:     interval,
+		Aggregations: make(map[string]Aggregation),
+	}
+}
+
+// WithMinDocCount the lower count threshold for a bucket to be included in the results
+func (d *DateHistogramAggregation) WithMinDocCount(minCount int64) *DateHistogramAggregation {
+	d.MinDocCount = minCount
+	return d
+}
+
+// WithTimeZone overwriting the default UTC timezone
+func (d *DateHistogramAggregation) WithTimeZone(tz string) *DateHistogramAggregation {
+	d.TimeZone = tz
+	return d
+}
+
+// AddOrder of the returned buckets
+func (d *DateHistogramAggregation) AddOrder(orders ...Order) *DateHistogramAggregation {
+	d.Order = append(d.Order, orders...)
+	return d
+}
+
+// AddSubAggregation to the TermsAggregation with the provided name
+func (d *DateHistogramAggregation) AddSubAggregation(name string, agg Aggregation) BucketAggregation {
+	if d.Aggregations == nil {
+		d.Aggregations = map[string]Aggregation{name: agg}
+	} else {
+		d.Aggregations[name] = agg
+	}
+
+	return d
+}
+
+// ConvertSubAggregations uses the provided converter to convert all the sub aggregations in this TermsAggregation
+func (d *DateHistogramAggregation) ConvertSubAggregations(converter AggregateVersionConverter) (map[string]Aggregation, error) {
+	convertedAggs := make(map[string]Aggregation, len(d.Aggregations))
+
+	for name, agg := range d.Aggregations {
+		cAgg, cErr := converter(agg)
```

Comment:
naming nit: just use `err` unless there is more than 1 error value in scope

Reply:
I've gotten in the habit of not using `err` to avoid wrong error issues I've seen pop up as bugs time and time again 😅 I can change it though if you don't like it.

Reply:
That's fair, but naming the variable for each check separately is very unidiomatic.  If the code is shadowing `err` and not immediately checking the result and returning then it should be refactored.